http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2ChannelHealthChecker.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal.http2;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils.warnIfNotInEventLoop;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.http2.DefaultHttp2PingFrame;
+import io.netty.handler.codec.http2.Http2PingFrame;
+import io.netty.util.concurrent.ScheduledFuture;
+import java.time.Duration;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * Used by {@link Http2PingedChannelPool} to periodically check the health of HTTP2 connections via PING frames.
+ *
+ * If a channel is found to be unhealthy, this will invoke {@link ChannelPipeline#fireExceptionCaught(Throwable)}.
+ */
+@SdkInternalApi
+class Http2ChannelHealthChecker {
+    private static final Logger log = Logger.loggerFor(Http2ChannelHealthChecker.class);
+    private static final Http2PingFrame DEFAULT_PING_FRAME = new DefaultHttp2PingFrame(0);
+
+    private final Channel channel;
+    private final long pingTimeoutMillis;
+
+    private ScheduledFuture<?> periodicPing;
+    private long lastPingSendTime = 0;
+    private long lastPingAckTime = Long.MAX_VALUE;
+
+    Http2ChannelHealthChecker(Channel channel, Duration pingTimeout) {
+        this.channel = channel;
+        this.pingTimeoutMillis = pingTimeout.toMillis();
+    }
+
+    public void start() {
+        warnIfNotInEventLoop(channel.eventLoop());
+
+        channel.pipeline().addLast(new PingResponseListener());
```

Comment:
Does it make sense to guard against adding more than 1?

Reply:
+1